remote-trait-object/src/service/serde_support.rs

```diff
@@ -0,0 +1,91 @@
+// Copyright 2020 Kodebox, Inc.
+// This file is part of CodeChain.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use super::*;
+use serde::de::{Deserialize, Deserializer};
+use serde::ser::{Serialize, Serializer};
+
+pub struct SArc<T: ?Sized + Service> {
+    value: std::cell::Cell<Option<Arc<T>>>,
+}
+
+impl<T: ?Sized + Service> SArc<T> {
+    pub fn new(value: Arc<T>) -> Self {
+        SArc {
+            value: std::cell::Cell::new(Some(value)),
+        }
+    }
+
+    pub(crate) fn take(&self) -> Arc<T> {
+        self.value.take().unwrap()
+    }
+
+    pub fn unwrap(self) -> Arc<T> {
+        self.value.take().unwrap()
+    }
+}
+
+/// This manages thread-local keys for port, which will be used in serialization of
+/// SArc. Note that this is required even in the inter-process setup.
+/// TODO: check that serde doens't spawn a thread while serializing.
+pub mod port_thread_local {
+    use super::*;
+    use std::cell::RefCell;
+
+    // TODO
+    // If a service call another service, this PORT setting might be stacked (at most twice).
```

Comment:
@majecty While some service object is handling incoming call (a.k.a `dispatch_and_call()`), it might involve another outgoing call depending on the implementation of actual method. In that case, `Handle::call()` will try to set port again on the same thread, where the already set port might be different one.

Reply:
OK, I understood.

Reply:
Btw I think there is nothing to be changed after the library becomes stabilized. Please ignore the last part of the comment.